README.md
# Analytics Statistics: Retrieve data from Google Analytics

[![Latest Version](https://img.shields.io/github/release/spatie/analytics-statistics.svg?style=flat-square)](https://github.com/spatie/analytics-statistics/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/spatie/analytics-statistics/master.svg?style=flat-square)](https://travis-ci.org/spatie/analytics-statistics)
[![Quality Score](https://img.shields.io/scrutinizer/g/spatie/analytics-statistics.svg?style=flat-square)](https://scrutinizer-ci.com/g/spatie/analytics-statistics)
[![Total Downloads](https://img.shields.io/packagist/dt/spatie/analytics-statistics.svg?style=flat-square)](https://packagist.org/packages/spatie/analytics-statistics)

An opinionated package to retrieve Google Analytics data. Compatible with PHP >= 5.3.

If you're using laravel 5, check out our [laravel-analytics](https://github.com/spatie/laravel-analytics/) package!

## Install

This package can be installed through Composer.

``` bash
composer require spatie/analytics-statistics
```

### How to obtain the credentials to communicate with Google Analytics

If you haven't already done so, [set up a Google Analytics property](https://support.google.com/analytics/answer/1042508) and [install the tracking code on your site](https://support.google.com/analytics/answer/1008080?hl=en#GA).

This package needs valid configuration values for `siteId`, `clientId` and `serviceEmail`. Additionally a `p12-file`is required.
 
To obtain these credentials start by going to the [Google Developers Console](https://console.developers.google.com).

If you don't have a project present in the console yet, create one.
If you click on the project name, you'll see a menu item `APIs` under `APIs & auth` on the left hand side. Click it to go the the Enabled API's screen. On that screen you should enable the Analytics API.
Now, again under the `APIs & Auth`-menu click `Credentials`.
On this screen you should press `Create new Client ID`. In the creation screen make sure you select application type `Service Account` and key type `P12-key`.

This wil generate a new public/private key pair and the .p12-file will get downloaded to your machine. Store this file in the location specified in the configfile of this package.

In the properties of the newly created Service Account you'll find the values for the `serviceEmail` and `clientId` listed as `CLIENT ID` and `EMAIL ADDRESS`.

To find the right value for `siteId` log in to [Google Analytics](http://www.google.be/intl/en/analytics/) go the the Admin section.
In the property-column select the website name of which you want to retrieve data, then click `View Settings` in the `View`-column.
The value presented as `View Id` prepended with 'ga:' can be used as `siteId`.

Make sure you've added the `ANALYTICS_SERVICE_EMAIL` to the Google Analytics Account otherwise you will get a `403: User does not have any Google Analytics Account` error. [You can read Google's instructions here](http://support.google.com/analytics/bin/answer.py?hl=en&answer=1009702).

If you want to use the realtime methods you should [request access](https://docs.google.com/forms/d/1qfRFysCikpgCMGqgF3yXdUyQW4xAlLyjKuOoOEFN2Uw/viewform) to the beta version of [Google's Real Time Reporting API](https://developers.google.com/analytics/devguides/reporting/realtime/v3/).

## Usage

### Setup

The easiest way to set up Analytics is via the factory method `\Spatie\Analytics\Analytics::create()`. The `create` method requires the following parameters:

```php
/**
 * @param  string $siteId
 * @param  string $clientId
 * @param  string $serviceEmail
 * @param  string $certificatePath
 */
```

You can optionally add some extra parameters to set up caching:

```php
/**
 * @param  \Spatie\Analytics\Cache $cache
 * @param  int $cacheLifetimeInMinutes
 * @param  int $realTimeCacheLifetime
 */
```

*Note: If you want to use a cache, you'll need to write your own implementation/adapter using the `\Spatie\Analytics\Analytics\Cache` interface.*

#### Example

```php
use Spatie\Analytics\Analytics;

$analytics = Analytics::create(
    'ga:xxxxxxxx',
    'xxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.apps.googleusercontent.com',
    'xxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx@developer.gserviceaccount.com',
    '/keys/analytics/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-privatekey.p12',
    $myCache,
    60 * 24 * 2,
    5
);
```

When the setup is done you can easily retrieve Analytics data. Methods generally return a plain array.

Here is an example to retrieve visitors and pageview data for the last seven days.

```php
/*
* $analyticsData now contains an array with 3 columns: "date", "visitors" and "pageViews"
*/
$analyticsData = $analytics->getVisitorsAndPageViews(7);
```

Here's another example to get the 20 most visited pages of the last 365 days

```php
/*
* $analyticsData now contains an array with 2 columns: "url" and "pageViews"
*/
$analyticsData = $analytics->getMostVisitedPages(365, 20);
```

## Provided methods

### Visitors and pageviews

These methods return an array with columns "date", "vistors" and "pageViews". When grouping by yearMonth, the first column will be called "yearMonth".

```php
/**
 * Get the amount of visitors and pageviews
 *
 * @param  int $numberOfDays
 * @param  string $groupBy Possible values: date, yearMonth
 * @return array
 */
public function getVisitorsAndPageViews($numberOfDays = 365, $groupBy = 'date')

/**
 * Get the amount of visitors and pageviews for the given period
 *
 * @param  \DateTime $startDate
 * @param  \DateTime $endDate
 * @param  string $groupBy  Possible values: date, yearMonth
 * @return array
 */
public function getVisitorsAndPageViewsForPeriod($startDate, $endDate, $groupBy = 'date')
```

### Keywords

These methods return an array with columns "keyword" and "sessions".

```php
   /**
 * Get the top keywords
 *
 * @param  int $numberOfDays
 * @param  int $maxResults
 * @return array
 */
public function getTopKeywords($numberOfDays = 365, $maxResults = 30)

/**
 * Get the top keywords for the given period
 *
 * @param  \DateTime $startDate
 * @param  \DateTime $endDate
 * @param  int $maxResults
 * @return array
 */
public function getTopKeyWordsForPeriod($startDate, $endDate, $maxResults = 30)
```

### Referrers

These methods return an array with columns "url" and "pageViews".

```php
/**
 * Get the top referrers
 *
 * @param  int $numberOfDays
 * @param  int $maxResults
 * @return array
 */
public function getTopReferrers($numberOfDays = 365, $maxResults = 20)

/**
 * Get the top referrers for the given period
 *
 * @param  \DateTime $startDate
 * @param  \DateTime $endDate
 * @param  $maxResults
 * @return array
 */
public function getTopReferrersForPeriod($startDate, $endDate, $maxResults)
```

### Browsers

These methods return an array with columns "browser" and "sessions".

If there are  more used browsers than the number specified in maxResults, then a new resultrow with browser-name "other" will be appended with a sum of all the remaining browsers.

```php
/**
 * Get the top browsers
 *
 * @param  int $numberOfDays
 * @param  int $maxResults
 * @return array
 */
public function getTopBrowsers($numberOfDays = 365, $maxResults = 6)

/**
 * Get the top browsers for the given period
 *
 * @param  \DateTime $startDate
 * @param  \DateTime $endDate
 * @param  $maxResults
 * @return array
 */
public function getTopBrowsersForPeriod($startDate, $endDate, $maxResults) 
```

### Most visited pages

These methods return an array with columns "url" and "pageViews".

```php
/**
 * Get the most visited pages
 *
 * @param  int $numberOfDays
 * @param  int $maxResults
 * @return array
 */
public function getMostVisitedPages($numberOfDays = 365, $maxResults = 20)

/**
 * Get the most visited pages for the given period
 *
 * @param  \DateTime $startDate
 * @param  \DateTime $endDate
 * @param  int $maxResults
 * @return array
 */
public function getMostVisitedPagesForPeriod($startDate, $endDate, $maxResults = 20)
```

### Currently active visitors

This method uses the [Real Time Reporting API](https://developers.google.com/analytics/devguides/reporting/realtime/v3/). It returns the amount of visitors that is viewing your site
right now.

```php
/**
 * Get the number of active users currently on the site
 * 
 * @param  array $others
 * @return array
 */
public function getActiveUsers($others = array())
```
   
### All other Google Analytics Queries

To perform all other GA queries use  ```performQuery```.  [Google's Core Reporting API](https://developers.google.com/analytics/devguides/reporting/core/v3/common-queries) provides more information on on which metrics and dimensions might be used. 

```php
/**
 * Call the query method on the autenthicated client
 *
 * @param  \DateTime $startDate
 * @param  \DateTime $endDate
 * @param  $metrics
 * @param  array $others
 * @return mixed
 */
public function performQuery($startDate, $endDate, $metrics, $others = array())
```

## Testing

Run the tests with:

```bash
vendor/bin/phpunit
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email freek@spatie.be instead of using the issue tracker.

## Credits

- [Freek Van der Herten](https://github.com/freekmurze)
- [Matthias De Winter](https://github.com/MatthiasDeWinter)
- [Sebastian De Deyne](https://github.com/sebastiandedeyne)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `spatie/analytics-statistics` will be documented in this file

##1.0.1 2015-06-02
Bugfix: `Analytics::getTopBrowsersForPeriod()` now works

##1.0 2015-06-01
First release!


tests/AnalyticsTest.php
<?php

namespace Spatie\Analytics\Test;

use DateTime;
use Mockery;
use Spatie\Analytics\Analytics;

class AnalyticsTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @var \Mockery\MockInterface  Mock of \Spatie\Analytics\GoogleClient
     */
    protected $client;

    /**
     * @var \Spatie\Analytics\Analytics
     */
    protected $analytics;

    /**
     * @var string
     */
    protected $siteId;

    /**
     * @var \DateTime
     */
    protected $now;

    public function setUp()
    {
        $this->client = Mockery::mock('\Spatie\Analytics\GoogleClient');
        $this->siteId = '12345';
        $this->now = new DateTime;

        $this->analytics = new Analytics($this->client, $this->siteId);
    }

    public function testGetVisitorsAndPageViews()
    {
        $startDate = $this->now->modify('-1 year')->format('Y-m-d');
        $endDate = date('Y-m-d');

        $this->client
            ->shouldReceive('performQuery')
            ->with($this->siteId, $startDate, $endDate, "ga:visits,ga:pageviews",
                array(
                    'dimensions' => 'ga:date'
                )
            )
            ->andReturn((object) array('rows' => array(array('20140101', 2, 3))));

        $googleResult = $this->analytics->getVisitorsAndPageViews();

        $resultProperties = array('date', 'visitors', 'pageViews');

        $this->assertTrue(count($googleResult) === 1);

        foreach ($resultProperties as $property) {
            $this->assertArrayHasKey($property, $googleResult[0]);
        }
    }

    public function testGetTopKeywords()
    {
        $startDate = $this->now->modify('-1 year')->format('Y-m-d');
        $endDate = date('Y-m-d');

        $this->client
            ->shouldReceive('performQuery')
            ->with($this->siteId, $startDate, $endDate, "ga:sessions",
                array(
                    'dimensions' => 'ga:keyword',
                    'sort' => '-ga:sessions',
                    'max-results' => 30,
                    'filters' => 'ga:keyword!=(not set);ga:keyword!=(not provided)'
                )
            )
            ->andReturn((object) array('rows' => array(array('first', 'second'))));

        $googleResult = $this->analytics->getTopKeyWords();

        $this->assertEquals($googleResult, array(array('keyword' => 'first', 'sessions' => 'second')));
    }

    public function testGetTopReferrers()
    {
        $startDate = $this->now->modify('-1 year')->format('Y-m-d');
        $endDate = date('Y-m-d');

        $this->client
            ->shouldReceive('performQuery')
            ->with($this->siteId, $startDate, $endDate, "ga:pageviews",
                array(
                    'dimensions' => 'ga:fullReferrer',
                    'sort' => '-ga:pageviews',
                    'max-results' => 20
                )
            )
            ->andReturn((object) array('rows' => array(array('foundUrl', '123'))));

        $googleResult = $this->analytics->getTopReferrers();

        $this->assertEquals($googleResult, array(array('url' => 'foundUrl', 'pageViews' => '123')));
    }

    public function testGetTopBrowsers()
    {
        $startDate = $this->now->modify('-1 year')->format('Y-m-d');
        $endDate = date('Y-m-d');

        $this->client
            ->shouldReceive('performQuery')
            ->with($this->siteId, $startDate, $endDate, "ga:sessions",
                array(
                    'dimensions' => 'ga:browser',
                    'sort' => '-ga:sessions'
                )
            )
            ->andReturn((object) array('rows' => array(array('Google Chrome', '123'))));

        $googleResult = $this->analytics->getTopBrowsers();

        $this->assertEquals($googleResult, array(array('browser' => 'Google Chrome', 'sessions' => '123')));
    }

    public function testGetMostVisitedPages()
    {
        $startDate = $this->now->modify('-1 year')->format('Y-m-d');
        $endDate = date('Y-m-d');

        $this->client
            ->shouldReceive('performQuery')
            ->with($this->siteId, $startDate, $endDate, "ga:pageviews",
                array(
                    'dimensions' => 'ga:pagePath',
                    'sort' => '-ga:pageviews',
                    'max-results' => 20
                )
            )
            ->andReturn((object) array('rows' => array(array('visited url', '123'))));

        $googleResult = $this->analytics->getMostVisitedPages();

        $this->assertEquals($googleResult, array(array('url' => 'visited url', 'pageViews' => '123')));
    }

    public function testGetSiteIdByUrl()
    {
        $testUrl = 'www.google.com';
        $siteId = 12345;

        $this->client->shouldReceive('getSiteIdByUrl')->with($testUrl)->andReturn($siteId);

        $result = $this->analytics->getSiteIdByUrl($testUrl);

        $this->assertEquals($result, $siteId);
    }

    public function testPerformQuery()
    {
        $startDate = $this->now->modify('-1 year');
        $endDate = new DateTime();
        $metrics = 'ga:somedummymetric';
        $others = array('first', 'second');

        $queryResult = 'result';

        $this->client
            ->shouldReceive('performQuery')
            ->with($this->siteId, $startDate->format('Y-m-d'), $endDate->format('Y-m-d'), $metrics, $others)
            ->andReturn($queryResult);

        $googleResult = $this->analytics->performQuery($startDate, $endDate, $metrics, $others);

        $this->assertSame($googleResult, $queryResult);
    }

    public function testIsEnabled()
    {
        $enabledAnalytics = new Analytics($this->client, $this->siteId);
        $this->assertTrue($enabledAnalytics->isEnabled());

        $disabledAnalytics = new Analytics($this->client);
        $this->assertFalse($disabledAnalytics->isEnabled());
    }
    
    public function testPerformRealTimeQuery()
    {
        $metrics = 'rt:somedummymetric';
        $others = array('first', 'second');

        $queryResult = 'result';

        $this->client
            ->shouldReceive('performRealTimeQuery')
            ->with($this->siteId, $metrics, $others)
            ->andReturn($queryResult);

        $googleResult = $this->analytics->performRealTimeQuery($metrics, $others);

        $this->assertSame($googleResult, $queryResult);
    }

    public function testGetActiveUsers()
    {
        $others = array('first', 'second');
        $metrics = 'rt:activeUsers';

        $this->client
                ->shouldReceive('performRealTimeQuery')
                ->with($this->siteId, $metrics, $others)
                ->andReturn((object) array('rows' => array(array(0, '500'))));

            $googleResult = $this->analytics->getActiveUsers($others);

            $this->assertInternalType('int', $googleResult);
    }
}


src/GoogleClient.php
<?php

namespace Spatie\Analytics;

use Carbon\Carbon;
use Exception;
use Google_Client;
use Google_Service_Analytics;

class GoogleClient
{
    /**
     * @var Google_Service_Analytics
     */
    protected $service;

    /**
     * @var \Spatie\Analytics\Cache
     */
    protected $cache;

    /**
     * @var int
     */
    protected $cacheLifeTimeInMinutes;

    /**
     * @var int
     */
    protected $realTimeCacheLifeTimeInSeconds;

    /**
     * @param  Google_Client $client
     * @param  \Spatie\Analytics\Cache $cache
     * @param  int $cacheLifeTimeInMinutes
     * @param  int $realTimeCacheLifeTimeInSeconds
     */
    public function __construct(
        Google_Client $client,
        Cache $cache = null,
        $cacheLifeTimeInMinutes = 0,
        $realTimeCacheLifeTimeInSeconds = 0
    ) {
        $this->service = new Google_Service_Analytics($client);
        $this->cache = $cache;

        $this->cacheLifeTimeInMinutes = $cacheLifeTimeInMinutes;
        $this->realTimeCacheLifeTimeInSeconds = $realTimeCacheLifeTimeInSeconds;
    }

    /**
     * Query the Google Analytics Service with given parameters.
     *
     * @param  int    $id
     * @param  string $startDate
     * @param  string $endDate
     * @param  string $metrics
     * @param  array  $others
     * @return mixed
     */
    public function performQuery($id, $startDate, $endDate, $metrics, array $others = array())
    {
        $cacheName = $this->determineCacheName(func_get_args());

        if ($this->useCache() && $this->cache->has($cacheName)) {
            return $this->cache->get($cacheName);
        }

        $googleAnswer = $this->service->data_ga->get($id, $startDate, $endDate, $metrics, $others);

        if ($this->useCache()) {
            $this->cache->put($cacheName, $googleAnswer, $this->cacheLifeTimeInMinutes);
        }

        return $googleAnswer;
    }
    
    /**
     * Query the Google Analytics Real Time Reporting Service with given parameters.
     *
     * @param  int    $id
     * @param  string $metrics
     * @param  array  $others
     * @return mixed
     */
    public function performRealTimeQuery($id, $metrics, array $others = array())
    {
        $realTimeCacheName = $this->determineRealTimeCacheName(func_get_args());

        if ($this->useRealTimeCache() && $this->cache->has($realTimeCacheName)) {
            return $this->cache->get($realTimeCacheName);
        }

        $googleAnswer =  $this->service->data_realtime->get($id, $metrics, $others);

        if ($this->useRealTimeCache()) {
            $this->cache->put(
                $realTimeCacheName,
                $googleAnswer,
                Carbon::now()->addSeconds($this->realTimeCacheLifeTimeInSeconds)
            );
        }

        return $googleAnswer;
    }

    /**
     * Get a site Id by its URL.
     *
     * @param  string $url
     * @return mixed
     * @throws \Exception
     */
    public function getSiteIdByUrl($url)
    {
        $siteIds = $this->getAllSiteIds();

        if (isset($siteIds[$url])) {
            return $siteIds[$url];
        }

        throw new Exception("Site $url is not present in your Analytics account.");
    }

    /**
     * Get all siteIds.
     *
     * @return array
     */
    public function getAllSiteIds()
    {
        static $siteIds = null;

        if (! is_null($siteIds)) {
            return $siteIds;
        }

        foreach ($this->service->management_profiles->listManagementProfiles("~all", "~all") as $site) {
            $siteIds[$site['websiteUrl']] = 'ga:'.$site['id'];
        }

        return $siteIds;
    }

    /**
     * Determine the cache name for the set of query properties given.
     *
     * @param  array $properties
     * @return string
     */
    protected function determineCacheName(array $properties)
    {
        return 'spatie.laravel-analytics.'.md5(serialize($properties));
    }

    /**
     * Determine if request to Google should be cached.
     *
     * @return bool
     */
    protected function useCache()
    {
        return isset($this->cache) && $this->cacheLifeTimeInMinutes > 0;
    }

    /**
     * Set the cache time.
     *
     * @param  int $cacheLifeTimeInMinutes
     * @return self
     */
    public function setCacheLifeTimeInMinutes($cacheLifeTimeInMinutes)
    {
        $this->cacheLifeTimeInMinutes = $cacheLifeTimeInMinutes;

        return $this;
    }

    /**
     * Determine the cache name for RealTime function calls for the set of query properties given.
     *
     * @param  array $properties
     * @return string
     */
    protected function determineRealTimeCacheName(array $properties)
    {
        return 'spatie.laravel-analytics.RealTime.'.md5(serialize($properties));
    }

    /**
     * Determine if RealTime request to Google should be cached.
     *
     * @return bool
     */
    protected function useRealTimeCache()
    {
        return isset($this->cache) && $this->realTimeCacheLifeTimeInSeconds > 0;
    }

    /**
     * Set the cache time.
     *
     * @param  int $realTimeCacheLifeTimeInSeconds
     * @return self
     */
    public function setRealTimeCacheLifeTimeInMinutes($realTimeCacheLifeTimeInSeconds)
    {
        $this->realTimeCacheLifeTimeInSeconds = $realTimeCacheLifeTimeInSeconds;

        return $this;
    }
}


src/Analytics.php
<?php

namespace Spatie\Analytics;

use Carbon\Carbon;
use DateTime;
use Exception;
use Google_Auth_AssertionCredentials;
use Google_Client;

class Analytics
{
    /**
     * @var \Spatie\Analytics\GoogleClient
     */
    protected $client;

    /**
     * @var string
     */
    protected $siteId;

    /**
     * @param  \Spatie\Analytics\GoogleClient $client
     * @param  string $siteId
     */
    public function __construct(GoogleClient $client, $siteId = '')
    {
        $this->client = $client;
        $this->siteId = $siteId;
    }

    /**
     * Get the amount of visitors and pageViews.
     *
     * @param  int $numberOfDays
     * @param  string $groupBy  Possible values: date, yearMonth
     * @return array
     */
    public function getVisitorsAndPageViews($numberOfDays = 365, $groupBy = 'date')
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getVisitorsAndPageViewsForPeriod($startDate, $endDate, $groupBy);
    }

    /**
     * Get the amount of visitors and pageviews for the given period.
     *
     * @param  \DateTime $startDate
     * @param  \DateTime $endDate
     * @param  string $groupBy  Possible values: date, yearMonth
     * @return array
     */
    public function getVisitorsAndPageViewsForPeriod(DateTime $startDate, DateTime $endDate, $groupBy = 'date')
    {
        $answer = $this->performQuery(
            $startDate,
            $endDate,
            'ga:visits,ga:pageviews',
            array('dimensions' => 'ga:'.$groupBy)
        );

        if (is_null($answer->rows)) {
            return array();
        }

        $visitorData = array();

        foreach ($answer->rows as $dateRow) {
            $visitorData[] = array(
                $groupBy    => Carbon::createFromFormat(($groupBy == 'yearMonth' ? 'Ym' : 'Ymd'), $dateRow[0]),
                'visitors'  => $dateRow[1],
                'pageViews' => $dateRow[2]
            );
        }

        return $visitorData;
    }

    /**
     * Get the top keywords.
     *
     * @param  int $numberOfDays
     * @param  int $maxResults
     * @return array
     */
    public function getTopKeywords($numberOfDays = 365, $maxResults = 30)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getTopKeyWordsForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the top keywords for the given period.
     *
     * @param  \DateTime $startDate
     * @param  \DateTime $endDate
     * @param  int $maxResults
     * @return array
     */
    public function getTopKeyWordsForPeriod(DateTime $startDate, DateTime $endDate, $maxResults = 30)
    {
        $answer = $this->performQuery(
            $startDate,
            $endDate,
            'ga:sessions',
            array(
                'dimensions' => 'ga:keyword',
                'sort' => '-ga:sessions',
                'max-results' => $maxResults,
                'filters' => 'ga:keyword!=(not set);ga:keyword!=(not provided)'
            )
        );

        if (is_null($answer->rows)) {
            return array();
        }

        $keywordData = array();

        foreach ($answer->rows as $pageRow) {
            $keywordData[] = array(
                'keyword' => $pageRow[0],
                'sessions' => $pageRow[1]
            );
        }

        return $keywordData;
    }

    /**
     * Get the top referrers.
     *
     * @param  int $numberOfDays
     * @param  int $maxResults
     * @return array
     */
    public function getTopReferrers($numberOfDays = 365, $maxResults = 20)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getTopReferrersForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the top referrers for the given period.
     *
     * @param  \DateTime $startDate
     * @param  \DateTime $endDate
     * @param  int $maxResults
     * @return array
     */
    public function getTopReferrersForPeriod(DateTime $startDate, DateTime $endDate, $maxResults)
    {
        $answer = $this->performQuery(
            $startDate,
            $endDate,
            'ga:pageviews',
            array(
                'dimensions' => 'ga:fullReferrer',
                'sort' => '-ga:pageviews',
                'max-results' => $maxResults
            )
        );

        if (is_null($answer->rows)) {
            return array();
        }

        $referrerData = array();

        foreach ($answer->rows as $pageRow) {
            $referrerData[] = array(
                'url' => $pageRow[0],
                'pageViews' => $pageRow[1]
            );
        }

        return $referrerData;
    }

    /**
     * Get the top browsers.
     *
     * @param  int $numberOfDays
     * @param  int $maxResults
     * @return array
     */
    public function getTopBrowsers($numberOfDays = 365, $maxResults = 6)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getTopBrowsersForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the top browsers for the given period.
     *
     * @param  \DateTime $startDate
     * @param  \DateTime $endDate
     * @param  int $maxResults
     * @return array
     */
    public function getTopBrowsersForPeriod(DateTime $startDate, DateTime $endDate, $maxResults)
    {
        $answer = $this->performQuery(
            $startDate,
            $endDate,
            'ga:sessions',
            array(
                'dimensions' => 'ga:browser',
                'sort' => '-ga:sessions'
            )
        );

        if (is_null($answer->rows)) {
            return array();
        }

        $browserData = array();

        foreach ($answer->rows as $browserRow) {
            $browserData[] = array(
                'browser' => $browserRow[0],
                'sessions' => $browserRow[1]
            );
        }

        $browserCollection = array_slice($browserData, 0, $maxResults - 1);

        if (count($browserData) > $maxResults) {
            $otherBrowsers = array_slice($browserData, $maxResults - 1);

            $sessions = array_map(function ($browser) {
                return $browser['sessions'];
            }, $otherBrowsers);

            $browserCollection[] = array(
                'browser' => 'other',
                'sessions' => array_sum($sessions)
            );
        }

        return $browserCollection;
    }

    /**
     * Get the most visited pages.
     *
     * @param  int $numberOfDays
     * @param  int $maxResults
     * @return array
     */
    public function getMostVisitedPages($numberOfDays = 365, $maxResults = 20)
    {
        list($startDate, $endDate) = $this->calculateNumberOfDays($numberOfDays);

        return $this->getMostVisitedPagesForPeriod($startDate, $endDate, $maxResults);
    }

    /**
     * Get the number of active users currently on the site
     *
     * @param  array $others
     * @return int
     */
    public function getActiveUsers($others = array())
    {
        $answer = $this->performRealTimeQuery('rt:activeUsers', $others);
    
        if (is_null($answer->rows)) {
            return 0;
        }
        
        return $answer->rows[0][0];
    }
    
    /**
     * Get the most visited pages for the given period.
     *
     * @param  \DateTime $startDate
     * @param  \DateTime $endDate
     * @param  int $maxResults
     * @return array
     */
    public function getMostVisitedPagesForPeriod(DateTime $startDate, DateTime $endDate, $maxResults = 20)
    {
        $answer = $this->performQuery(
            $startDate,
            $endDate,
            'ga:pageviews',
            array(
                'dimensions' => 'ga:pagePath',
                'sort' => '-ga:pageviews',
                'max-results' => $maxResults
            )
        );

        if (is_null($answer->rows)) {
            return array();
        }

        $pagesData = array();

        foreach ($answer->rows as $pageRow) {
            $pagesData[] = array(
                'url' => $pageRow[0],
                'pageViews' => $pageRow[1]
            );
        }

        return $pagesData;
    }

    /**
     * Returns the site id (ga:xxxxxxx) for the given url.
     *
     * @param  string $url
     * @return string
     */
    public function getSiteIdByUrl($url)
    {
        return $this->client->getSiteIdByUrl($url);
    }

    /**
     * Call the query method on the authenticated client.
     *
     * @param  \DateTime $startDate
     * @param  \DateTime $endDate
     * @param  string $metrics
     * @param  array $others
     * @return mixed
     */
    public function performQuery(DateTime $startDate, DateTime $endDate, $metrics, $others = array())
    {
        return $this->client->performQuery(
            $this->siteId,
            $startDate->format('Y-m-d'),
            $endDate->format('Y-m-d'),
            $metrics,
            $others
        );
    }

    /**
     * Call the real time query method on the authenticated client.
     *
     * @param  string $metrics
     * @param  array $others
     * @return mixed
     */
    public function performRealTimeQuery($metrics, $others = array())
    {
        return $this->client->performRealTimeQuery(
            $this->siteId,
            $metrics,
            $others
        );
    }
    
    /**
     * Return true if this site is configured to use Google Analytics.
     *
     * @return bool
     */
    public function isEnabled()
    {
        return $this->siteId != '';
    }

    /**
     * Returns an array with the current date and the date minus the number of days specified.
     *
     * @param  int $numberOfDays
     * @return array
     */
    protected function calculateNumberOfDays($numberOfDays)
    {
        $endDate = Carbon::today();
        $startDate = Carbon::today()->subDays($numberOfDays);

        return array(
            $startDate,
            $endDate
        );
    }

    /**
     * Create a new instance via a set of parameters
     * 
     * @param  string $siteId
     *         Ex. ga:xxxxxxxx
     * @param  string $clientId
     *         Ex. xxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.apps.googleusercontent.com
     * @param  string $serviceEmail
     *         Ex. xxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx@developer.gserviceaccount.com
     * @param  string $certificatePath
     *         Ex. /../keys/analytics/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx-privatekey.p12
     * 
     * @param  \Spatie\Analytics\Cache|null $cache
     * @param  int $cacheLifetimeInMinutes
     * @param  int $realTimeCacheLifetime
     * 
     * @return \Spatie\Analytics\Analytics
     * 
     * @throws \Exception
     */
    public static function create($siteId, $clientId, $serviceEmail, $certificatePath, Cache $cache = null,
        $cacheLifetimeInMinutes = 0, $realTimeCacheLifetime = 0
    ) {
        if (! file_exists($certificatePath)) {
            throw new Exception("Can't find the .p12 certificate in: $certificatePath");
        }

        $client = new Google_Client(
            array(
                'oauth2_client_id' => $clientId,
                'use_objects'      => true,
            )
        );

        $client->setAccessType('offline');

        $client->setAssertionCredentials(
            new Google_Auth_AssertionCredentials(
                $serviceEmail,
                array('https://www.googleapis.com/auth/analytics.readonly'),
                file_get_contents($certificatePath)
            )
        );

        $googleApi = new GoogleClient($client, $cache);

        $googleApi        
            ->setCacheLifeTimeInMinutes($cacheLifetimeInMinutes)
            ->setRealTimeCacheLifeTimeInMinutes($realTimeCacheLifetime);

        return new static($googleApi, $siteId);
    }
}
